tools/smidump/src/tools/smilint.1.in
.\"
.\" $Id: smilint.1.in,v 1.15 2002/11/12 10:03:42 schoenw Exp $
.\"
.TH smilint 1  "June 21, 2002" "IBR" "SMI Tools"
.SH NAME
smilint \- syntax and semantic checks of SMIv1/v2, SPPI, and SMIng modules
.SH SYNOPSIS
.B smilint
[
.B "-Vhersm"
] [
.BI "-c " file
] [
.BI "-p " module
] [
.BI "-l " level
] [
.BI "-i " error-pattern
]
.I "module(s)"
.SH DESCRIPTION
The \fBsmilint\fP program is used to check MIB or PIB modules for syntax
errors and semantics at some degree.  SMIv1/v2 and SMIng style
MIB modules as well as SPPI PIB modules are supported.
.PP
The rules that \fBsmilint\fP is based on are taken from RFC 1155, RFC
1212 and RFC 1215 for SMIv1, RFCs 2578-2580 for SMIv2, RFC 3159 for SPPI,
and the specification of SMIng that has been a research project
within the Network Management Research Group of the IRTF.
.SH OPTIONS
.TP
\fB-V, --version\fP
Show the smilint version and exit.
.TP
\fB-h, --help\fP
Show a help text and exit.
.TP
\fB-e, --error-list\fP
Show a list of all known error messages and exit. Error messages can
have associated tags, shown in braces at the end of each line. The
tags can be used with the -i option to ignore certain error messages.
.TP
\fB-r, --recursive\fP
Report errors and warnings also for recursively imported modules. 
.TP
\fB-s, --severity\fP
Show the error severity in brackets before error messages.
.TP
\fB-m, --error-names\fP
Show the error names in braces before error messages.
.TP
\fB-c \fIfile\fB, --config=\fIfile\fP
Read \fIfile\fP instead of any other (global and user)
configuration file.
.TP
\fB-p \fImodule\fB, --preload=\fImodule\fP
Preload the module \fImodule\fP before reading the main
module(s). This may be helpful if an incomplete main module misses to
import some definitions.
.TP
\fB-l \fIlevel\fB, --level=\fIlevel\fP
Report errors and warnings up to the given severity \fIlevel\fP.
A higher level shows more warnings. The actual severity levels are:
.sp
.nf
 0   internal error, no recovery possible,
 1   major SMI/SPPI error, recovery somehow possible but
     may lead to severe problems,
 2   SMI/SPPI error, probably tolerated by some
     implementations,
 3   SMI/SPPI error, tolerated by many implementations,
 4   something which is recommended to be changed,
 5   something that might be ok, but might not be
     recommended under some circumstances,
 6   just a notice.
.fi
.sp
Higher levels are currently not used and lead to the same effects as
level 6 does. Note that errors up to level 3 are errors violating the
specifications and must be fixed by the responsible author. Level 3 is
the default level.
.TP
\fB-i \fIprefix\fB, --ignore=\fIprefix\fP
Ignore all errors that have a tag which matches \fIprefix\fP.
A list of error tags can be retrieved by calling smilint with the
-e option.
.TP
.I module(s)
These are the modules to be checked. If a module argument represents a
path name (identified by containing at least one dot or slash character),
this is assumed to be the exact file to read. Otherwise, if a module is
identified by its plain module name, it is searched according to libsmi
internal rules. See \fBsmi_config(3)\fP for more details.
.SH "EXAMPLE"
This example checks the file RMON2-MIB in the current directory (note
that the `./' prefix ensures this). The error level is raised to 6
and warnings that claim about identifier names that exceed a length
of 32 characters are suppressed.
.nf

  $ smilint -l 6 -i namelength-32 ./RMON2-MIB
  ./RMON2-MIB:3935: unexpected type restriction
  ./RMON2-MIB:3936: unexpected type restriction
  ./RMON2-MIB:3937: unexpected type restriction
  ./RMON2-MIB:3938: unexpected type restriction
  ./RMON2-MIB:3939: unexpected type restriction
  ./RMON2-MIB:3940: unexpected type restriction
  ./RMON2-MIB:4164: scalar object must not have a `read-create' access value

.fi
.SH "SEE ALSO"
The 
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/projects/libsmi/" "."
Other commonly used MIB checkers are
.BR mosy "(1) and "
.BR smicng "(1)."
.SH "AUTHOR"
(C) 1999-2002 Frank Strauss, TU Braunschweig, Germany <strauss@ibr.cs.tu-bs.de>
.br
(C) 1999-2002 J. Schoenwaelder, TU Braunschweig, Germany <schoenw@ibr.cs.tu-bs.de>
.br
(C) 2001-2002 T. Klie, TU Braunschweig, Germany <tklie@ibr.cs.tu-bs.de>
.br
(C) 2002 M. Bunkus, TU Braunschweig, Germany <bunkus@ibr.cs.tu-bs.de>
.br
and contributions by many other people.
.br


tools/smidump/src/tools/smilint.c
/*
 * smilint.c --
 *
 *      MIB module checker main program.
 *
 * Copyright (c) 1999 Frank Strauss, Technical University of Braunschweig.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smilint.c,v 1.43 2002/03/05 16:19:59 strauss Exp $
 */

#include <config.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "shhopt.h"



/*
 * These are functions that are not officially exported by the libsmi.
 * See the original prototype definitions in lib/error.h.
 */

extern int smiGetErrorSeverity(int id);
extern char* smiGetErrorTag(int id);
extern char* smiGetErrorMsg(int id);



static int mFlag = 0;	/* show the name for error messages */
static int sFlag = 0;	/* show the severity for error messages */
static int eFlag = 0;	/* print the list of possible error messages */
static int flags;


typedef struct Error {
    int id;
    int severity;
    char *tag;
    char *msg;
} Error;



static int compare(const void *v1, const void *v2)
{
    Error *err1 = (Error *) v1;
    Error *err2 = (Error *) v2;

    if (err1->severity < err2->severity) {
	return -1;
    }
    if (err1->severity > err2->severity) {
	return 1;
    }
    return strcmp(err1->msg, err2->msg);
}



static void errors()
{
    int i;
    size_t cnt;
    Error *errors;
    
    for (cnt = 0, i = 0; smiGetErrorSeverity(i) >= 0; i++) {
	cnt++;
    }
    
    errors = malloc(cnt * sizeof(Error));
    if (! errors) {
	fprintf(stderr, "smilint: malloc failed - running out of memory\n");
	exit(1);
    }
    memset(errors, 0, cnt * sizeof(Error));
    for (i = 0; i < (int)cnt; i++) {
	errors[i].id = i;
	errors[i].severity = smiGetErrorSeverity(i);
	errors[i].tag = smiGetErrorTag(i);
	errors[i].msg = smiGetErrorMsg(i);
    }

    qsort(errors, cnt, sizeof(Error), compare);
    
    for (i = 0; i < (int)cnt; i++) {
	if (sFlag) {
	    printf("[%d] %s (%s)\n",
		   errors[i].severity, errors[i].msg, errors[i].tag);
	} else {
	    printf("%s (%s)\n", errors[i].msg, errors[i].tag);
	}
    }
    
    free(errors);
}



static void usage()
{
    fprintf(stderr,
	    "Usage: smilint [options] [module or path ...]\n"
	    "  -V, --version        show version and license information\n"
	    "  -h, --help           show usage information\n"
	    "  -c, --config=file    load a specific configuration file\n"
	    "  -p, --preload=module preload <module>\n"
	    "  -e, --error-list     print list of known error messages\n"
	    "  -m, --error-names    print the name of errors in braces\n"
	    "  -s, --severity       print the severity of errors in brackets\n"
	    "  -r, --recursive      print errors also for imported modules\n"
	    "  -l, --level=level    set maximum level of errors and warnings\n"
	    "  -i, --ignore=prefix  ignore errors matching prefix pattern\n");
}



static void help() { usage(); exit(0); }
static void version() { printf("smilint " SMI_VERSION_STRING "\n"); exit(0); }
static void config(char *filename) { smiReadConfig(filename, "smilint"); }
static void preload(char *module) { smiLoadModule(module); }
static void recursive() { flags |= SMI_FLAG_RECURSIVE; smiSetFlags(flags); }
static void level(int lev) { smiSetErrorLevel(lev); }
static void ignore(char *ign) { smiSetSeverity(ign, 9999); }



static void
errorHandler(char *path, int line, int severity, char *msg, char *tag)
{
    if (path) {
	fprintf(stderr, "%s:%d: ", path, line);
    }
    if (sFlag) {
	fprintf(stderr, "[%d] ", severity);
    }
    if (mFlag) {
	fprintf(stderr, "{%s} ", tag);
    }
    fprintf(stderr, "%s\n", msg);

    if (severity <= 0) {
	exit(1);
    }
}



int main(int argc, char *argv[])
{
    int i;

    static optStruct opt[] = {
	/* short long              type        var/func       special       */
	{ 'h', "help",           OPT_FLAG,   help,          OPT_CALLFUNC },
	{ 'V', "version",        OPT_FLAG,   version,       OPT_CALLFUNC },
	{ 'c', "config",         OPT_STRING, config,        OPT_CALLFUNC },
	{ 'p', "preload",        OPT_STRING, preload,       OPT_CALLFUNC },
	{ 'e', "error-list",     OPT_FLAG,   &eFlag,        0 },
	{ 'm', "error-names",    OPT_FLAG,   &mFlag,        0 },
	{ 's', "severity",       OPT_FLAG,   &sFlag,        0 },
	{ 'r', "recursive",      OPT_FLAG,   recursive,     OPT_CALLFUNC },
	{ 'l', "level",          OPT_INT,    level,         OPT_CALLFUNC },
	{ 'i', "ignore",         OPT_STRING, ignore,        OPT_CALLFUNC },
	{ 0, 0, OPT_END, 0, 0 }  /* no more options */
    };
    
    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-c") == argv[i]) ||
	    (strstr(argv[i], "--config") == argv[i])) break;
    if (i == argc) 
	smiInit("smilint");
    else
	smiInit(NULL);

    flags = smiGetFlags();
    flags |= SMI_FLAG_ERRORS;
    flags |= SMI_FLAG_NODESCR;
    smiSetFlags(flags);

    optParseOptions(&argc, argv, opt, 0);

    if (sFlag || mFlag) {
	smiSetErrorHandler(errorHandler);
    }

    if (eFlag) {
	errors();
	return 0;
    }
    
    for (i = 1; i < argc; i++) {
	if (smiLoadModule(argv[i]) == NULL) {
	    fprintf(stderr, "smilint: cannot locate module `%s'\n",
		    argv[i]);
	    smiExit();
	    exit(1);
	}
    }

    smiExit();

    return 0;
}
